### PR TITLE
Preserve `encoding` attr on `PackOp` for `TileAndDistributeToWorkgroups`

### PR DESCRIPTION
**Pull request by bjacob**
During `TileAndDistributeToWorkgroups`, in the case of dynamic tile sizes, the `LowerDispatchWorkgroupCountFromSetEncodingOp` pattern needs to call `materializeEncodingValueFn`, passing it a tensor type, expecting that tensor type to have a `TensorEncodingAttr`. The problem is that `MaterializeEncodingPass` has already run, rewriting the `SetEncoding` op and its result tensor, which used to hold the `TensorEncodingAttr`, into a `PackOp` op, whose new result tensor does not anymore have a `TensorEncodingAttr`.

As a work-around for that, we make `MaterializeEncoding` preserve the `TensorEncodingAttr` as an attribute on the `PackOp` itself, so `TileAndDistributeToWorkgroups` can read it and reconstruct a `tensorTypeWithEncoding`, so `LowerDispatchWorkgroupCountFromSetEncodingOp` can call `materializeEncodingValueFn`.

### CODE REVIEW

**Comment by benvanik** [comment]
This is also a workaround worth filing an issue for and tagging for cleanup - needing to dynamically add attributes to ops is a big red flag that the op may need to be split (one before/after the materialize encodings pass) as it introduces a lot of ambiguity into the lowering sequence (the pass will behave differently when run instead of only being valid to be run based on the IR).

**Review by benvanik — APPROVED** [review]

**Comment by bjacob** [comment]
Filed #11608.

**Review by hanhanW — DISMISSED** [review]
Hey @bjacob, I need to take a deeper look into what's happening. I will unblock the landing after it.

**Comment by hanhanW** [comment]
I think I should review it with https://github.com/iree-org/iree/pull/11605

If I understand correctly, we need the encoding because we're going to create `vmvx.query_tile_sizes encoding(encoding)` op in TileAndDistribute. Does it mean that we want to defer the decision of the tiling sizes to execution time? Or we just want to defer it to later VMVX compilation time (or say before injecting microkernel calls?)?

The other question that I just have, are the information of inner_dims_pos and outer_dim_perms enough for microkernels?

(I'm looking for if we can get rid of attributes/encodings after materialization. Maybe encoding is a placeholder for the proper fix, so it's fine.)

**Comment by bjacob** [comment]
> I think I should review it with #11605

Up to you! I thought separate commits helped separate concerns, but it's still possible to treat them as two commits on #11605.

> 
> If I understand correctly, we need the encoding because we're going to create `vmvx.query_tile_sizes encoding(encoding)` op in TileAndDistribute. Does it mean that we want to defer the decision of the tiling sizes to execution time? Or we just want to defer it to later VMVX compilation time (or say before injecting microkernel calls?)?

We do want to defer deciding tile sizes to runtime, but the reason for that is not any such detail. So let us first discuss why we want deferring the tile size decision to runtime, and then separately we will discuss why we need the encoding here.

**Why defer tile size decision to runtime:**
* **TLDR:** Because it's what users really want, particularly in the VMVX backend. Simple compilation command lines creating universal bytecode modules and still leverage whichever CPU instructions the IREE runtime has support for in its ukernels.
* **Details:**
** ***Preamble.*** As you know, the new `--iree-flow-enable-data-tiling` command-line switch does not anymore take options describing target information. In LLVMCPU compilation command lines, that is now solely passed in the standard way, namely `--iree-llvm-target-triple` and `--iree-llvm-target-cpu-features`. That's great, but as I started work on migrating the VMVX+ukernels path to `--iree-flow-enable-data-tiling`, I realized that those flags are specific to the LLVMCPU backend, and don't exist with the VMVX backend. That is understandable, as the VMVX backend is all about creating universal portable bytecode modules. But, I thought, we want matmuls to be fast, and we can only achieve that if we know target information, so... something has to give... 
** ***An abandoned approach.*** At that time, the best compromise that I could think of was to create "multi" variants of the above LLVMCPU target flags, allowing to create VMVX modules that, although technically universal and portable, would have multiple `HAL::ExecutableVariant` for different tile sizes, which would happen to run faster on the corresponding targets. The VMVX loader would then automatically select the right variant based on the runtime host CPU information. This was proposed in #11377. @benvanik rejected the approach, but also provided the better idea...
** ***Enlightenment.*** Ben suggested: in the case of the VMVX backend, where we highly value portability of bytecode modules and are OK to compromise a bit on performance in exchange for that, it just makes more sense to use dynamic tile sizes, and let a ukernel do runtime tile size decisions. That way we have optimally simple compilation command lines --- no target information needs to be passed on the command line, unlike the above abandoned #11377. And the resulting bytecode modules are fully portable, not specialized for any target in any way, and still will use optimized tile sizes and kernels for whichever target variant we happen to have ukernels for in the IREE runtime. This is clearly a win for users, and should make VMVX+ukernels the backend that "just works" (and actually uses whichever features your CPU happens to have).

**Why we need the encoding here.**
* **TLDR:** This is just a quirk of the current design, contrary to the above question which was high-level user-centric. This is explained in this comment: https://github.com/iree-org/iree/pull/11603/files#diff-a22007f739c269b76bf20767ad05b2c03862fc22df8fdd341a692e06e6cf40dcR355 . It boils down to a discrepancy between:
** On the one hand, when `MaterializeEncoding` rewrites a `SetEncodingOp` into a `PackOp`, it drops the information that was contained in the `TensorEncoding`. The idea is that that information lives on in the materialized `PackOp` params (i.e. the `MaterializeEncodingInfo`). But that materialized information is not always containing all the information that was contained in the original `TensorEncoding`, as in the case of dynamic tile sizes, the `MaterializeEncodingInfo::innerTileSizes` are just `kDynamic`, i.e. almost no information at all. So from the perspective of something that is trying to call a `MaterializeEncodingValueFn` to obtain runtime values of dynamic tile sizes, the `PackOp` parameters, or equivalently the `MaterializeEncodingInfo` is not enough information. Still, this should not be a problem at all if **nothing else** than `MaterializeEncoding` needed to know `TensorEncoding`. And one would think that to be the case, since, in the case of dynamic shapes, after `MaterializeEncoding` has run, the `packOp`'s operand dimensions contain the actual runtime size Values.
** On the other hand, during `TileAndDistributeToWorkgroups`, the `LowerDispatchWorkgroupCountFromSetEncodingOp` needs to know the sizes of the pack op's outer dimensions... AND, as far as I understand, it can't access the pack op as a SSA value, it just doesn't live in the same scope, and attempts to do so have resulted in `UNKNOWN SSA VALUE`. So it has to re-compute the pack op's outer dimensions based on whichever information is enough to deduce it, and can be passed to it. The system implemented between this PR and #11605 is the best that I could think of.

But clearly, all this is only needed due to this specific combination of constraints in the current specific design.

> 
> The other question that I just have, are the information of inner_dims_pos and outer_dim_perms enough for microkernels?

Microkernels need to know all of the pack op's parameters, including the tile sizes. You can find reference code for the `pack` microkernel [here](https://github.com/iree-org/iree/blob/2315a906f4c1e7c4bcf1572b6938e2284e2c2e2b/runtime/src/iree/builtins/ukernel/tools/pack_test.cc#L19-L65). Here, the tile sizes appear as `out_size2`, `out_size3`. It will make it more concrete how it uses that information. Then in optimized code, this information is [used to dispatch](https://github.com/iree-org/iree/blob/2315a906f4c1e7c4bcf1572b6938e2284e2c2e2b/runtime/src/iree/builtins/ukernel/arch/arm_64/pack_arm_64.c#L11-L32) to a specific optimized inner loop.

> 
> (I'm looking for if we can get rid of attributes/encodings after materialization. Maybe encoding is a placeholder for the proper fix, so it's fine.)

I agree that's the interesting question here! And we've touched on exactly that in the above answer to "Why we need the encoding here". Philosophically we shouldn't need encodings past materialization.

**Comment by bjacob** [comment]
@hanhanW let me know by direct message:
> Feel free to land the PR that Ben approved. Overall they look good to me!
So I will click 'dismiss review'.
